backend/src/spotify.py

```diff
@@ -0,0 +1,166 @@
+import os
+from typing import List
+import spotipy
+from spotipy.oauth2 import SpotifyOAuth
+from src.album import Album
+
+from src.text_formatting import format_ms_as_mins_and_secs
+
+scope = [
+    "user-library-read",
+    "user-read-currently-playing",
+    "user-read-playback-state",
+    "playlist-modify-public",
+    "playlist-modify-private",
+    "playlist-read-private",
+    "playlist-read-collaborative",
+    "playlist-modify-private",
+    "playlist-modify-public",
+]
+
+
+class PlaybackInfo:
+    def __init__(
+        self,
+        track_title,
+        album_title,
+        track_artists,
+        album_artists,
+        artwork_url,
+        track_progress,
+        track_duration,
+        album_progress,
+        album_duration,
+    ):
+        self.track_title = track_title
+        self.album_title = album_title
+        self.track_artists = track_artists
+        self.album_artists = album_artists
+        self.artwork_url = artwork_url
+        self.track_progress = track_progress
+        self.track_duration = track_duration
+        self.album_progress = album_progress
+        self.album_duration = album_duration
+
+    def get_formatted_artists(self):
+        return ", ".join(self.track_artists)
+
+    def get_formatted_track_progress(self):
+        return (
+            format_ms_as_mins_and_secs(self.track_progress)
+            + " / "
+            + format_ms_as_mins_and_secs(self.track_duration)
+        )
+
+    def get_formatted_album_progress(self):
+        return (
+            format_ms_as_mins_and_secs(self.album_progress)
+            + " / "
+            + format_ms_as_mins_and_secs(self.album_duration)
+            + f" ({(100*self.album_progress/self.album_duration):02.0f}%)"
+        )
+
+
+class SpotifyClient:
+    def __init__(self):
+        client_id = os.getenv("SPOTIFY_CLIENT_ID")
+        client_secret = os.getenv("SPOTIFY_SECRET")
+        redirect_uri = os.getenv("SPOTIPY_REDIRECT_URI")
+
+        self.client = spotipy.Spotify(
+            auth_manager=SpotifyOAuth(
```

Comment:
Works really nicely, but I'm quite surprised by this library behaviour if I'm understanding so might be something to chat about at some point - I feel like if we're using the server-server route to authenticate with Spotify, then we may as well use the [ClientCredentials flow](https://spotipy.readthedocs.io/en/2.22.1/#client-credentials-flow) over OAuth:
* For example, if we ran this on the cloud, how do you authorise the app with spotify the first time? (locally this triggers a firefox window to pop up)
* That said, presumably our real goal would be to eventually get multiple users each seeing & managing their own playlists, which probably wants a slightly different design (i.e. use OAuth, but getting them to take part in it as part of the front-end flow, rather than it being baked into our backend flow (if I'm understanding correctly)

Reply:
Yeah, I'll admit, I'm still not entirely clear on the OAuth flow here. Although it seems necessary over ClientCredentials for users to be able to manage their own playlists and libraries. I know we'll be looking more into this sort of stuff in Module 11, but I'll try and get it figured out and cleaner by then 😄. Definitely would be one to have a chat about in a 1-1 soon.

Reply:
Yeah, I suspect the issue is that Spotipy wants to handle it on the backend entirely, but we probably need to link it in with the frontend a bit more - definitely one we can discuss and explore